projects/charger/src/begin_sequence.c

```diff
@@ -0,0 +1,71 @@
+#include "begin_sequence.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can.h"
+#include "can_transmit.h"
+#include "charger_controller.h"
+#include "charger_events.h"
+#include "control_pilot.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "gpio.h"
+#include "status.h"
+
+static const GpioAddress s_cp_select = CONTROL_PILOT_SEL_PIN;
+static const GpioAddress s_charger_sense = CHARGER_SENSE_PIN;
+static const GpioAddress s_relay_en = RELAY_EN_PIN;
+static const GpioAddress s_load_sw_en = LOAD_SW_EN_PIN;
+
+StatusCode prv_sequence(const CanMessage *msg, void *context, CanAckStatus *ack_reply) {
+  GpioState state = NUM_GPIO_STATES;
+
+  // 1. Ensure charger is on
+  gpio_get_state(&s_charger_sense, &state);
+  if (state != GPIO_STATE_HIGH) {
+    CAN_TRANSMIT_CHARGER_FAULT(EE_CHARGER_FAULT_CHARGER_OFF);
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+
+  // 2. Get control pilot PWM reading
+  uint16_t cp_pwm_ret = control_pilot_get_current();
+
+  // 3. Set relay and load switch state
+  gpio_set_state(&s_relay_en, GPIO_STATE_HIGH);
+  gpio_set_state(&s_load_sw_en, GPIO_STATE_HIGH);
+
+  // 4. Enable charging via control pilot
+  gpio_set_state(&s_cp_select, GPIO_STATE_HIGH);
+
+  // 5. Activate charger
+  charger_controller_activate(cp_pwm_ret);
+
+  return STATUS_CODE_OK;
+}
+
+void begin_sequence_process_event(const Event *e) {
+  if (e->id == CHARGER_CHARGE_EVENT_BEGIN) {
+    CAN_TRANSMIT_REQUEST_TO_CHARGE();
+  }
+}
+
+StatusCode begin_sequence_init() {
+  GpioSettings settings = {
+    .direction = GPIO_DIR_OUT,       //
+    .state = GPIO_STATE_LOW,         //
+    .resistor = GPIO_RES_NONE,       //
+    .alt_function = GPIO_ALTFN_NONE  //
+  };
+  gpio_init_pin(&s_cp_select, &settings);
+  gpio_init_pin(&s_relay_en, &settings);
+  gpio_init_pin(&s_load_sw_en, &settings);
+
+  settings.direction = GPIO_DIR_IN;
+  gpio_init_pin(&s_charger_sense, &settings);
+
+  can_register_rx_handler(SYSTEM_CAN_MESSAGE_ALLOW_CHARGING, prv_sequence, NULL);
+  control_pilot_init();
```

Comment:
Should this be initialized in `main.c`? (Just seems a bit strange - what if another module wants to use control_pilot later?)

Reply:
This is to demonstrate that the `begin_sequence` module owns the `control_pilot` module. This pattern is used elsewhere as well (e.g. centre console, prooobably other projects too), I think it would make sense if I drew the block diagram but I'm too lazy

Reply:
Fair enough, I might add a comment to `control_pilot.h` to that effect but that's up to you.

Reply:
Added comment to `control_pilot.h`